src/Perspex.Controls/GridSplitter.cs

```diff
@@ -2,42 +2,240 @@
-        /// Initializes a new instance of the <see cref="GridSplitter"/> class.
+        /// Gets or sets the orientation of the GridsSlitter, if null, it's inferred from column/row defenition(should be auto).
         /// </summary>
-        public GridSplitter()
+        public Orientation? Orientation { get { return GetValue(OrientationProperty); } set { SetValue(OrientationProperty, value); } }
+
+
+        static GridSplitter()
         {
-            Cursor = new Cursor(StandardCursorType.SizeWestEast);
+            PseudoClass(OrientationProperty, o => o == Perspex.Controls.Orientation.Vertical, ":vertical");
+            PseudoClass(OrientationProperty, o => o == Perspex.Controls.Orientation.Horizontal, ":horizontal");
         }
 
+
         protected override void OnDragDelta(VectorEventArgs e)
         {
-            int col = GetValue(Grid.ColumnProperty);
+            _resizer.DragDelta(e);
+        }
 
-            if (_grid != null && col > 0)
+        /// <summary>
+        /// If orientation is not set, method automatically calculates orientation based column/row auto size
+        /// </summary>
+        /// <returns></returns>
+        private void AutoSetOrientation()
+        {
+            if (Orientation.HasValue)
             {
-                var size = _grid.ColumnDefinitions[col - 1].ActualWidth + e.Vector.X;               
-                _grid.ColumnDefinitions[col - 1].Width = new GridLength(
-                    Math.Max(0, size),
-                    GridUnitType.Pixel);
+                return;
             }
+            if (_grid.ColumnDefinitions[GetValue(Grid.ColumnProperty)].Width.IsAuto)
+            {
+                Orientation = Perspex.Controls.Orientation.Vertical;
+                return;
+            }
+            if (_grid.RowDefinitions[GetValue(Grid.RowProperty)].Height.IsAuto)
+            {
+                Orientation = Perspex.Controls.Orientation.Horizontal;
+                return;
+            }
+            throw new InvalidOperationException("GridSpliter Should have Orientation, width or height set.");
         }
 
         protected override void OnAttachedToVisualTree(VisualTreeAttachmentEventArgs e)
         {
-            base.OnAttachedToVisualTree(e);
             _grid = this.GetVisualParent<Grid>();
+            AutoSetOrientation();
+            switch (Orientation)
+            {
+                case Perspex.Controls.Orientation.Vertical:
+                    _resizer = new VerticalColumnsResizer(_grid, GetValue(Grid.ColumnProperty));
+                    break;
+                case Perspex.Controls.Orientation.Horizontal:
+                    _resizer = new HorizontalGridColumnsResizer(_grid, GetValue(Grid.RowProperty));
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException();
+            }
+            base.OnAttachedToVisualTree(e);
+        }
+    }
+
+    internal interface IGridColumnsResizer
```

Comment:
I'm a little concerned that this is over-engineered. The definitions for the interface, abstract class, and 2 derived classes far outweigh the size of the rest of the control. I suspect that the code can be cut drastically just by using `if`s rather than interfaces and abstract methods etc.

It's fine to be extensible when there can be extensions, but here you have precisely 2 options, so it seems like overkill to me.